wormhole-connect/src/utils/tokens.ts

```diff
@@ -0,0 +1,95 @@
+import { createUmi } from '@metaplex-foundation/umi-bundle-defaults';
+import {
+  fetchDigitalAsset,
+  mplTokenMetadata,
+} from '@metaplex-foundation/mpl-token-metadata';
+import { publicKey } from '@metaplex-foundation/umi';
+import {
+  Wormhole,
+  PlatformToChains,
+  TokenId,
+  chainToPlatform,
+  Network,
+  NativeAddress,
+} from '@wormhole-foundation/sdk';
+import { getWormholeContextV2 } from 'config';
+import { Contract } from 'ethers';
+
+interface TokenMetadataFromRpc {
+  symbol: string;
+  name: string;
+  icon?: string;
+}
+
+export async function getTokenMetadataFromRpc(
+  tokenId: TokenId,
+): Promise<TokenMetadataFromRpc | undefined> {
+  const wh = await getWormholeContextV2();
+  const platform = chainToPlatform(tokenId.chain);
+
+  switch (platform) {
+    case 'Solana':
+      return getTokenMetadataSolana(wh, tokenId);
+    case 'Evm':
+      return getTokenMetadataEvm(wh, tokenId);
+  }
+}
+
+export async function getTokenMetadataSolana(
+  wh: Wormhole<Network>,
+  tokenId: TokenId,
+): Promise<TokenMetadataFromRpc | undefined> {
+  try {
+    const platform = wh.getPlatform('Solana');
+    const rpc = platform.getRpc('Solana');
+    const umi = createUmi(rpc);
+    umi.use(mplTokenMetadata());
+
+    const { metadata } = await fetchDigitalAsset(
+      umi,
+      publicKey((tokenId.address as NativeAddress<'Solana'>).unwrap()),
+    );
+
+    const { symbol, name } = metadata;
+    let icon: string | undefined = undefined;
+
+    if (metadata.uri) {
+      // Try to get an image from ipfs
+      try {
+        const { image } = await (await fetch(metadata.uri)).json();
+        if (image) {
+          icon = image;
+        }
+      } catch (_) {
+        // Oh well we tried
+      }
+    }
+
+    return { symbol, name, icon };
+  } catch (e) {
+    return undefined;
+  }
+}
+
+export async function getTokenMetadataEvm(
+  wh: Wormhole<Network>,
+  tokenId: TokenId,
+): Promise<TokenMetadataFromRpc | undefined> {
+  try {
+    const ABI = [
+      'function symbol() view returns (string)',
+      'function name() view returns (string)',
+    ];
+
+    const platform = wh.getPlatform('Evm');
+    const rpc = platform.getRpc(tokenId.chain as PlatformToChains<'Evm'>);
+
+    const contract = new Contract(tokenId.address.toString(), ABI, rpc);
+    const symbol = await contract.symbol();
```

Comment:
We could do `Promise.all([contract.symbol(), contract.name()])` here

Reply:
https://github.com/wormhole-foundation/wormhole-connect/pull/2985/commits/cd7a756ed88a1adff5301769481951e06c6d5e18